test/Test/Db.hs

```diff
@@ -1,2 +1,79 @@
-module Test.Db where
-    
\ No newline at end of file
+module Test.Db 
+       ( dbSpecs
+       ) where
+
+import IW.App (AppEnv)
+import IW.Core.Id (Id (..))
+import IW.Core.Issue (Issue (..))
+import IW.Core.Repo (RepoName (..), RepoOwner (..))
+import IW.Core.SqlArray (SqlArray (..))
+import IW.Db (getIssues, prepareDb, upsertIssues)
+import IW.Effects.Log (runAppLogIO_)
+
+import Test.Assert (equals, equals_)
+import Test.Common
+import Test.Hspec (Spec, before_, describe, it)
+
+
+dbSpecs :: AppEnv -> Spec
+dbSpecs = joinSpecs "DB"
+    [ getIssuesSpec
+    , upsertIssuesSpec
+    ]
+
+getIssuesSpec :: AppEnv -> Spec
+getIssuesSpec env = describe "getIssues function" $
+    it "should return a list of length 1" $
+        env & (length <$> getIssues) `equals` 1
+
+upsertIssuesSpec :: AppEnv -> Spec
+upsertIssuesSpec env = before_ (runAppLogIO_ env prepareDb) $
+    describe "upsertIssues function" $ do
+        it "should leave the issues table unaffected" $
+            env & ( do before <- getIssues
+                       upsertIssues [invalidIssue]
+                       return before    
+                  ) `equals_` getIssues 
+        it "should insert an issue if its repo exists" $
+            env & ( do before <- getIssues
+                       upsertIssues [validIssue]
```

Comment:
Hmm.. I think it was an oversight from my side to remove `getIssueById` function. Current tests are actually not robust enough. It's not enough to check that the number of issues, but also the values inside the issue. We want to make sure that during insert we insert all relevant fields and during update we update all necessary fields. So the test actually should look like this:

```haskell
it "..." $ do
    env & issuesAfterUpsertInserting `returnsSame` prec . length <$> getIssues
    env & getIssueById (Id 2) `equals` validIssue
```

Reply:
Got it.

Reply:
Question about this test. If the database automatically assigns the `id` how can we be sure that it will have the same `id` as when the issue was inserted? For example, the `WithId Issue` could have `withIdId` of `Id 2`, but when it gets inserted to DB it will change depending on its order.

Reply:
@rashadg1030 Good question. PostgreSQL has `RETURNING` for the insert. So you can return the `Id` of just inserted issue. So instead of `executeMany` in `insertIssues` you need to use `returning`:

https://github.com/kowainik/issue-wanted/blob/5b0d70f1886e8ce452f13769757ca73e3bf66718/src/IW/Db/Functions.hs#L81-L86

And, of course, you also need to use `RETURNING` inside SQL query.

P.S. For some reason `returning` doesn't have documentation... Would be nice to add it!

Reply:
Got it! I will look into `returning` and document my findings.

Reply:
I've added a function `issuesRowDifference` that solves the problem of just relying on the number of rows to test upsert functions.

```
-- | Returns difference between rows of the issue table after and before an action
issuesRowDifference :: App a -> App [Issue]
issuesRowDifference = beforeAfter getIssues (\\)
```
If `newRows ++ rowsBefore == afterRows` then `afterRows \\ beforeRows == newRows`.

Reply:
So do we still want to use `getIssueById` and `returning`? We could have both if that helps.

Reply:
Do we still need to use `returning`? Or was that just for the test?